test/dynamo/test_functions.py

```diff
@@ -1248,6 +1249,72 @@ def func():
         self.assertEqual(cnts.frame_count, 1)
         self.assertEqual(cnts.op_count, 1)
 
+    def test_dataclass_factory(self):
+        @dataclass
+        class Output:
+            scalar: int = 2
+            named_tensors: Dict[str, torch.Tensor] = field(default_factory=dict)
+            lists: List[torch.Tensor] = field(default_factory=list)
+
+            def scale(self):
+                return self.scalar * 2
+
+        def fn(x):
+            # Check default dict assignment
+            a = Output(1)
+            # Check that dataclass methods can be inlined
+            scaled_value = a.scale()
+
+            # Check that normal assignment works
+            b = Output(5, named_tensors={"x": x})
+
+            # Check default int assignment
+            c = Output()
+
+            # Check that the default members are properly initialized
+            if isinstance(a.named_tensors, dict):
+                x = torch.sin(x)
+
+            # Change dataclass
+            c.scalar = 6
+            c.named_tensors["x"] = x
+
+            # Return dataclaass as well to check reconstruction
+            return c, torch.cos(x) * scaled_value + b.named_tensors["x"] + c.scalar
```

Comment:
@jansel the test returns a modified dataclass object and compares them later to validate the reconstruction logic.